### PR TITLE
Change to using argparse.

### PR DESCRIPTION
**Pull request by RichWellum**
Move to using argparse.

Help looks like:

rwellum@RWELLUM-M-34DF:[~/Desktop/Boxes]: iosxrv-x64-vbox/iosxr_iso2vbox.py iosxrv-fullk9-x64.iso -o -v -h
usage: iosxr_iso2vbox.py [-h] [-a [ARTIFACTORY]] [-o] [-v] ISO_FILE

A tool to create an IOS XRv Vagrant VirtualBox box from an IOS XRv ISO.

The ISO will be installed, booted, configured and unit-tested.
"vagrant ssh" provides access to IOS XR Linux global-vrf namespace
with internet access.

positional arguments:
  ISO_FILE              local ISO filename or remote URI ISO filename...

optional arguments:
  -h, --help            show this help message and exit
  -a [ARTIFACTORY], --artifactory [ARTIFACTORY]
                        add optional reason for uploading this box
  -o, --create_ova      additionally use vboxmange to export an OVA
  -v, --verbose         turn on verbose messages

E.g.: iosxr-xrv64-vbox/iosxr_iso2vbox.py iosxrv-fullk9-x64.iso -a 'new image' -o -v

Tested it of course and flake8 is good.

Just don't like the "--artifactory [ARTIFACTORY]" - as that's a bit ugly.

### CODE REVIEW

File: iosxr_iso2vbox.py [hunk 1]
```diff
@@ -95,43 +97,53 @@ def main(argv):
...
+        formatter_class=RawTextHelpFormatter,
+        description='A tool to create an IOS XRv Vagrant VirtualBox box from ' +
+        'an IOS XRv ISO.\n' '\nThe ISO will be installed, booted, configured ' +
+        'and unit-tested. \n"vagrant ssh" provides access ' +
+        'to IOS XR Linux global-vrf namespace \nwith internet access.',
+        epilog="E.g.: iosxr-xrv64-vbox/iosxr_iso2vbox.py " +
+        "iosxrv-fullk9-x64.iso -a 'new image' -o -v")
+    parser.add_argument('ISO_FILE',
+                        help='local ISO filename or remote URI ISO filename...')
+    parser.add_argument('-a', '--artifactory', nargs='?',
```
Comment (RichWellum) [inline iosxr_iso2vbox.py]: Suggestions to avoid: --artifactory [ARTIFACTORY] would be appreciated. @glennmatthews

Comment (glennmatthews) [inline iosxr_iso2vbox.py]: `metavar='REASON'`

File: iosxr_iso2vbox.py [hunk 2]
```diff
@@ -95,43 +97,53 @@ def main(argv):
...
-            copy_to_artifactory = True
-            reason = arg
-            if not reason:
-                reason = 'No reason for update specified'
-        elif opt in ('-v', '--verbose'):
-            verbose = True
-        elif opt in ('-o', '--ova'):
-            create_ova = True
+    parser = argparse.ArgumentParser(
+        formatter_class=RawTextHelpFormatter,
```
Comment (glennmatthews) [inline iosxr_iso2vbox.py]: Unless you also want to manually wrap the argument help strings, you probably want `RawDescriptionHelpFormatter`, which preserves whitespace in description and epilog only.

File: iosxr_iso2vbox.py [hunk 3]
```diff
@@ -95,43 +97,53 @@ def main(argv):
...
+        description='A tool to create an IOS XRv Vagrant VirtualBox box from ' +
+        'an IOS XRv ISO.\n' '\nThe ISO will be installed, booted, configured ' +
+        'and unit-tested. \n"vagrant ssh" provides access ' +
+        'to IOS XR Linux global-vrf namespace \nwith internet access.',
+        epilog="E.g.: iosxr-xrv64-vbox/iosxr_iso2vbox.py " +
+        "iosxrv-fullk9-x64.iso -a 'new image' -o -v")
+    parser.add_argument('ISO_FILE',
+                        help='local ISO filename or remote URI ISO filename...')
+    parser.add_argument('-a', '--artifactory', nargs='?',
+                        const='check_string_for_empty',
```
Comment (glennmatthews) [inline iosxr_iso2vbox.py]: Why not `const='No reason for update specified'` to save you the need to check below?

File: iosxr_iso2vbox.py [hunk 4]
```diff
@@ -95,43 +97,53 @@ def main(argv):
...
+                        help='add optional reason for uploading this box')
```
Comment (glennmatthews) [inline iosxr_iso2vbox.py]: `help="Upload box to Artifactory. You can optionally specify a reason for uploading this box."`

File: iosxr_iso2vbox.py [hunk 5]
```diff
@@ -95,43 +97,53 @@ def main(argv):
...
+                        help='turn on verbose messages')
+
+    args = parser.parse_args()
+
+    # Handle Input ISO (Local or URI)
+    if re.search(':/', args.ISO_FILE):
+        # URI Image
+        cmdstring = 'scp %s@%s .' % (getpass.getuser(), args.ISO_FILE)
+        print('==> Will attempt to scp the remote image to current working dir. You may be required to enter your password.')
+        time.sleep(2)
```
Comment (glennmatthews) [inline iosxr_iso2vbox.py]: Why?

File: iosxr_iso2vbox.py [hunk 6]
```diff
@@ -95,43 +97,53 @@ def main(argv):
...
-        elif opt in ('-o', '--ova'):
-            create_ova = True
+    parser = argparse.ArgumentParser(
+        formatter_class=RawTextHelpFormatter,
+        description='A tool to create an IOS XRv Vagrant VirtualBox box from ' +
+        'an IOS XRv ISO.\n' '\nThe ISO will be installed, booted, configured ' +
+        'and unit-tested. \n"vagrant ssh" provides access ' +
+        'to IOS XR Linux global-vrf namespace \nwith internet access.',
+        epilog="E.g.: iosxr-xrv64-vbox/iosxr_iso2vbox.py " +
+        "iosxrv-fullk9-x64.iso -a 'new image' -o -v")
```
Comment (glennmatthews) [inline iosxr_iso2vbox.py]: Would be nice to provide examples of:
- box build with local iso
- box build with remote iso
- box build with ova export and upload to artifactory

File: iosxr_iso2vbox.py [hunk 7]
```diff
@@ -95,43 +97,53 @@ def main(argv):
...
+    parser.add_argument('ISO_FILE',
+                        help='local ISO filename or remote URI ISO filename...')
+    parser.add_argument('-a', '--artifactory', nargs='?',
+                        const='check_string_for_empty',
+                        help='add optional reason for uploading this box')
+    parser.add_argument('-o', '--create_ova', action='store_true',
+                        help='additionally use vboxmange to export an OVA')
```
Comment (glennmatthews) [inline iosxr_iso2vbox.py]: s/vboxmange/vboxmanage/ ?

File: iosxr_store_box.py [hunk 8]
```diff
@@ -33,25 +33,8 @@
...
-    '''
-    argv = command_string.split()
-    if debug is True:
-        print ('argv: %s' % argv)
-    result = subprocess.check_output(argv)
-    return result
+from iosxr_iso2vbox import run
 
 
 def main(argv):
```
Comment (glennmatthews) [inline iosxr_store_box.py:40]: Are you planning to refactor this one to use argparse in a future PR?

File: iosxr_store_box.py [hunk 9]
```diff
@@ -69,15 +52,19 @@ def main(argv):
 
     if artifactory_username is None:
         print('==> Please set ARTIFACTORY_USERNAME in your environment')
+        print('E.g. export "ARTIFACTORY_USERNAME=<username>"')
         sys.exit()
```
Comment (glennmatthews) [inline iosxr_store_box.py]: consider:

``` python
sys.exit("==> Please set ARTIFACTORY_USERNAME in your environment\n"
         "E.g., 'export ARTIFACTORY_USERNAME=<username>'")
```

File: iosxr_iso2vbox.py (hunk 2 above)
Comment (RichWellum) [inline iosxr_iso2vbox.py]: Ack.

File: iosxr_iso2vbox.py (hunk 6 above)
Comment (RichWellum) [inline iosxr_iso2vbox.py]: Done.

Is it appropriate in your opinion to do this in the epilog. Looks nice to me, but not sure if this a standard way to do it.

File: iosxr_iso2vbox.py (hunk 1 above)
Comment (RichWellum) [inline iosxr_iso2vbox.py]: Drat you told me this already - sorry forgot!!

Works well.

File: iosxr_iso2vbox.py (hunk 3 above)
Comment (RichWellum) [inline iosxr_iso2vbox.py]: Why not. :)

File: iosxr_iso2vbox.py (hunk 4 above)
Comment (RichWellum) [inline iosxr_iso2vbox.py]: Ack.

File: iosxr_iso2vbox.py (hunk 7 above)
Comment (RichWellum) [inline iosxr_iso2vbox.py]: Ack.

File: iosxr_iso2vbox.py (hunk 5 above)
Comment (RichWellum) [inline iosxr_iso2vbox.py]: Why the sleep? Legacy I think - have removed it.

File: iosxr_store_box.py (hunk 8 above)
Comment (RichWellum) [inline iosxr_store_box.py:40]: Ah, I did iosxrv_test.py correctly, but not this one. Yes:

US117901 - iosxrv_store.py - needs to be converted to use argparse.

File: iosxr_store_box.py (hunk 9 above)
Comment (RichWellum) [inline iosxr_store_box.py]: Yup and applied throughout.

File: iosxr_iso2vbox.py [hunk 10]
```diff
@@ -95,43 +97,53 @@ def main(argv):
...
+
+    # Handle Input ISO (Local or URI)
+    if re.search(':/', args.ISO_FILE):
+        # URI Image
+        cmdstring = 'scp %s@%s .' % (getpass.getuser(), args.ISO_FILE)
+        print('==> Will attempt to scp the remote image to current working dir. You may be required to enter your password.')
+        print('==> %s\n' % cmdstring)
+        subprocess.call(cmdstring, shell=True)
+        input_iso = os.path.basename(args.ISO_FILE)
+    else:
```
Comment (RichWellum) [inline iosxr_iso2vbox.py:132]: @glennmatthews Updated with your CR comments addressed.

File: iosxr_iso2vbox.py [hunk 11]
```diff
@@ -95,43 +97,53 @@ def main(argv):
...
-        elif opt in ('-o', '--ova'):
-            create_ova = True
-        else:
-            print(master_opts)
-            sys.exit()
+    # Handle artifactory
+    if args.artifactory is not None:
+        copy_to_artifactory = True
+        if args.artifactory is not 'No reason for update specified':
+            artifactory_reason = args.artifactory
```
Comment (glennmatthews) [inline iosxr_iso2vbox.py:140]: else, artifactory_reason == ''? Is that correct or should artifactory_reason be 'No reason for update specified'?

File: iosxr_iso2vbox.py (hunk 6 above)
Comment (glennmatthews) [inline iosxr_iso2vbox.py]: Yup - usage examples in the epilog is totally appropriate IMHO and IME.

**Comment by glennmatthews** [comment]
👍

**Pull request merged**